test/e2e/cluster_size_autoscaling.go

```diff
@@ -605,3 +652,117 @@ func setMigSizes(sizes map[string]int) {
+	By(fmt.Sprintf("Taint node %s", node.Name))
+	freshNode, err := c.Core().Nodes().Get(node.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	for _, taint := range freshNode.Spec.Taints {
+		if taint.Key == disabledTaint {
+			return nil
+		}
+	}
+	freshNode.Spec.Taints = append(freshNode.Spec.Taints, v1.Taint{
+		Key:    disabledTaint,
+		Value:  "DisabledForTest",
+		Effect: v1.TaintEffectNoSchedule,
+	})
+	_, err = c.Core().Nodes().Update(freshNode)
+	return err
+}
+
+func enableNode(c clientset.Interface, node *v1.Node) error {
+	By(fmt.Sprintf("Remove taint from node %s", node.Name))
+	freshNode, err := c.Core().Nodes().Get(node.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	newTaints := make([]v1.Taint, 0)
+	for _, taint := range freshNode.Spec.Taints {
+		if taint.Key != disabledTaint {
+			newTaints = append(newTaints, taint)
+		}
+	}
+
+	if len(newTaints) != len(freshNode.Spec.Taints) {
+		freshNode.Spec.Taints = newTaints
+		_, err = c.Core().Nodes().Update(freshNode)
+		return err
+	}
+	return nil
+}
+
+func runReplicatedPodOnEachNode(f *framework.Framework, nodes []v1.Node, id string, labels map[string]string) error {
+	By("Run a pod on each node")
+	for _, node := range nodes {
+		err := disableNode(f.ClientSet, &node)
+
+		defer func(n v1.Node) {
+			enableNode(f.ClientSet, &n)
+		}(node)
+
+		if err != nil {
+			return err
+		}
+	}
+	config := &testutils.RCConfig{
+		Client:         f.ClientSet,
+		InternalClient: f.InternalClientset,
+		Name:           id,
+		Namespace:      f.Namespace.Name,
+		Timeout:        defaultTimeout,
+		Image:          framework.GetPauseImageName(f.ClientSet),
+		Replicas:       0,
+		Labels:         labels,
+	}
+	err := framework.RunRC(*config)
+	if err != nil {
+		return err
+	}
+	rc, err := f.ClientSet.Core().ReplicationControllers(f.Namespace.Name).Get(id, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	for i, node := range nodes {
+		err = enableNode(f.ClientSet, &node)
+		if err != nil {
+			return err
+		}
+		*rc.Spec.Replicas = int32(i + 1)
+		rc, err = f.ClientSet.Core().ReplicationControllers(f.Namespace.Name).Update(rc)
+		if err != nil {
+			return err
+		}
+		err = wait.PollImmediate(5*time.Second, podTimeout, func() (bool, error) {
+			rc, err = f.ClientSet.Core().ReplicationControllers(f.Namespace.Name).Get(id, metav1.GetOptions{})
+			if err != nil || rc.Status.ReadyReplicas < int32(i+1) {
+				return false, nil
+			}
+			return true, nil
+		})
+		if err != nil {
+			return fmt.Errorf("failed to coerce RC into spawning a pod on node %s within timeout", node.Name)
+		}
+		err = disableNode(f.ClientSet, &node)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func manuallyIncreaseClusterSize(f *framework.Framework, originalSizes map[string]int) int {
+	By("Manually increase cluster size")
+	increasedSize := 0
+	newSizes := make(map[string]int)
+	for key, val := range originalSizes {
+		newSizes[key] = val + 2
```

Comment:
It was +2 before I moved this to a separate function and the number makes sense (I don't think it should be a parameter, as I don't see a usecase of changing it), but I made it a const to avoid magic numbers in code